### ISSUE TITLE
UntypedActor.onReceive should be able to throw Throwable

### ISSUE THREAD

**Issue by jroper** [akka/akka#20231]
`Throwable` and `Exception` are used inconsistently as the root exception type in Akka, and this causes problems in Java.

In one concrete use case that I have, I have an actor that subscribes to a stream, using `Sink.actorRef` to direct the messages to itself.  I'm also using the `BackoffSupervisor` pattern, so that when the stream fails, the supervisor will simply restart me, and I'll reconnect to the stream.  To implement this, when the stream fails, `Sink.actorRef` will send a `Status.Failure`, which has `cause` as a `Throwable`, and I need to rethrow that throwable to kill my actor.  Now, if I was using Scala, it would look like this:

```scala
case Status.Failure(cause) => throw cause
```

But in Java, I can't do that, since `onReceive` throws `Exception`, so I can't throw `cause()`, I have to wrap it.  There are a number of other cases in APIs in Akka where the type of exception that is passed is `Throwable`, but you can never directly throw this from an actor in Java because `onReceive` throws `Exception`.

Hence I think `UntypedActor.onReceive` should throw `Throwable`, not `Exception`, to be consistent with exception passing in the rest of Akka.

**Comment by ktoso** [comment]
Interesting observation, and a sound one I think.
Since these methods are a bit special – no-one is supposed to call them except us, I think we may get away with just fixing it in a compatible way hm...

I'm not sure if it's binary compatible, we'll need to check, but source compatible such change is surely:

```scala
@throws(classOf[Exception]) def onReceive(message: Any): Unit
// =>
@throws(classOf[Throwable]) def onReceive(message: Any): Unit
```

```java
// public void onReceive(Object message) throws Exception { 
// after change => 
public void onReceive(Object message) throws Exception { // still compiles
public void onReceive(Object message) throws Throwable { // is possible to write
```

**Comment by jroper** [comment]
To my knowledge, modifying checked exceptions on method signatures is only a source compatibility issue, it doesn't break binary compatibility, exception checking is the sole responsibility of the compiler, the runtime doesn't care.

**Comment by ktoso** [comment]
Will be solved in 2.4.8, coming this week
